libcanard/canard.h

```diff
@@ -1,528 +1,587 @@
+    ///
+    /// A simple and robust way of managing transfer-ID counters is to keep a separate static variable per subject-ID
+    /// and per (service-ID, server-node-ID) pair.
+    CanardTransferID transfer_id;
+
+    /// This is the actual transfer payload.
+    /// If the payload is empty (payload_size = 0), the payload pointer may be NULL.
+    /// The const pointer makes it incompatible with memory deallocation function, this is due to the limitations of C;
+    /// therefore, when freeing the memory allocated for the payload, cast away the pointer's const qualifier.
+    size_t      payload_size;
+    const void* payload;
+} CanardTransfer;
+
+/// Transfer subscription state. The application can register its interest in a particular kind of data exchanged
+/// over the bus by creating such subscription objects. Frames that carry data for which there is no active
+/// subscription will be silently dropped by the library.
+///
+/// WARNING: SUBSCRIPTION INSTANCES SHALL NOT BE COPIED OR MUTATED BY THE APPLICATION.
+///
+/// Every field is named starting with an underscore to emphasize that the application shall not modify it.
+/// Unfortunately, C, being such a limited language, does not allow us to construct a better API.
+///
+/// The memory footprint of a subscription is large. On a 32-bit platform it slightly exceeds half a KiB.
+/// This is an intentional time-memory trade-off: use a large look-up table to ensure predictable temporal properties.
+typedef struct CanardRxSubscription
+{
+    struct CanardRxSubscription* _next;  ///< Internal use only.
+
+    /// The current architecture is an acceptable middle ground between worst-case execution time and memory
+    /// consumption. Instead of statically pre-allocating a dedicated RX session for each remote node-ID here in
+    /// this table, we only keep pointers, which are NULL by default, populating a new RX session dynamically
+    /// on an ad-hoc basis when we first receive a transfer from that node. This is deterministic because our memory
+    /// allocation routines are assumed to be deterministic and we make at most one allocation per remote node,
+    /// but the disadvantage is that these additional operations lift the upper bound on the execution time.
+    /// Further, the pointers here add an extra indirection, which is bad for systems that leverage cached memory,
+    /// plus a pointer itself takes about 2-8 bytes of memory, too.
+    ///
+    /// A far more predictable and a much simpler approach is to pre-allocate states here statically instead of keeping
+    /// just pointers, but it would push the size of this instance from about 0.5 KiB to ~3 KiB for a typical 32-bit
+    /// system. Since this is a general-purpose library, we have to pick a middle ground so we use the more complex
+    /// but more memory-efficient approach.
+    struct CanardInternalRxSession* _sessions[CANARD_NODE_ID_MAX + 1U];
```

Comment:
> add a CANARD_NODE_COUNT_MAX.

I'm pretty sure it will backfire terribly when vendors of low-cost equipment start releasing hardware with CANARD_NODE_COUNT_MAX=16 to squeeze the software into underpowered MCUs. Let's not do it.

> I'd implement a sparse array for this

I thought about this, too, but I couldn't find a solution that would be both constant-complexity and sufficiently simple to implement. Do you have one in mind?

Reply:
If we don't add CANARD_NODE_COUNT_MAX then a sparse array provides value.

Reply:
I'm sure it does but I don't see how to implement it without sacrificing the temporal predictability. I think 512 (256 for AVR) bytes per subscription is bad, but it's not a disaster either.